src/layers/ordered.rs

```diff
@@ -0,0 +1,514 @@
+    pub vals: L,
+}
+
+// TODO: by-value merge
+impl<K, L, O> Add<Self> for OrderedLayer<K, L, O>
+where
+    K: Ord + Clone,
+    L: Trie,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        self.merge(&rhs)
+    }
+}
+
+impl<K, L, O> AddAssignByRef for OrderedLayer<K, L, O>
+where
+    K: Ord + Clone,
+    L: Trie,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    fn add_assign_by_ref(&mut self, other: &Self) {
+        *self = self.merge(other);
+    }
+}
+
+impl<K, L, O> AddByRef for OrderedLayer<K, L, O>
+where
+    K: Ord + Clone,
+    L: Trie,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    fn add_by_ref(&self, rhs: &Self) -> Self {
+        self.merge(rhs)
+    }
+}
+
+impl<K, L, O> NumEntries for OrderedLayer<K, L, O>
+where
+    K: Ord + Clone,
+    L: Trie,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    fn num_entries_shallow(&self) -> usize {
+        self.keys()
+    }
+
+    fn num_entries_deep(&self) -> usize {
+        self.tuples()
+    }
+
+    fn const_num_entries() -> Option<usize> {
+        None
+    }
+}
+
+impl<K, L, O> SharedRef for OrderedLayer<K, L, O>
+where
+    K: Ord + Clone,
+    L: Clone,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    type Target = Self;
+
+    fn try_into_owned(self) -> Result<Self::Target, Self> {
+        Ok(self)
+    }
+}
+
+impl<K, L, O> Trie for OrderedLayer<K, L, O>
+where
+    K: Ord + Clone,
+    L: Trie,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    type Key = K;
+    type ChildKey = L::Key;
+    type Item = (K, L::Item);
+    type Cursor = OrderedCursor<K, L, O>;
+    type MergeBuilder = OrderedBuilder<K, L::MergeBuilder, O>;
+    type TupleBuilder = UnorderedBuilder<K, L::TupleBuilder, O>;
+
+    fn keys(&self) -> usize {
+        self.keys.len()
+    }
+    fn tuples(&self) -> usize {
+        self.vals.tuples()
+    }
+    fn cursor_from(&self, lower: usize, upper: usize) -> Self::Cursor {
+        if lower < upper {
+            OrderedCursor {
+                bounds: (lower, upper),
+                pos: lower,
+                _types: PhantomData,
+            }
+        } else {
+            OrderedCursor {
+                bounds: (0, 0),
+                pos: 0,
+                _types: PhantomData,
+            }
+        }
+    }
+}
+
+/// Assembles a layer of this
+pub struct OrderedBuilder<K, L, O = usize>
+where
+    K: Ord,
+    O: OrdOffset,
+    <O as TryFrom<usize>>::Error: Debug,
+    <O as TryInto<usize>>::Error: Debug,
+{
+    /// Keys
+    pub keys: Vec<K>,
+    /// Offsets
+    pub offs: Vec<O>,
```

Comment:
I'd prefer this just being `offsets` instead of an abbreviated form. The docs on these fields aren't particular helpful either, "Keys to what? Offsets to what? Next layer of what?", they don't explain why the data's important